cpp/src/parquet/stream_reader.cc

```diff
@@ -202,23 +216,108 @@ void StreamReader::NextRowGroup() {
       column_readers_[i] = row_group_reader_->Column(i);
     }
     if (column_readers_[0]->HasNext()) {
+      row_group_row_offset_ = current_row_;
       return;
     }
   }
   // No more row groups found.
+  SetEof();
+}
+
+void StreamReader::SetEof() {
+  // Do not reset file_metadata_ to ensure queries on the number of
+  // rows/columns still function.
   eof_ = true;
   file_reader_.reset();
-  file_metadata_.reset();
   row_group_reader_.reset();
   column_readers_.clear();
   nodes_.clear();
 }
 
+int64_t StreamReader::SkipRows(int64_t num_rows_to_skip) {
+  if (0 != column_index_) {
+    throw ParquetException("Must finish reading current row before skipping rows.");
+  }
+  int64_t num_rows_remaining_to_skip = num_rows_to_skip;
+
+  while (!eof_ && (num_rows_remaining_to_skip > 0)) {
+    int64_t num_rows_in_row_group = row_group_reader_->metadata()->num_rows();
+    int64_t num_rows_remaining_in_row_group =
+        num_rows_in_row_group - current_row_ - row_group_row_offset_;
+
+    if (num_rows_remaining_in_row_group > num_rows_remaining_to_skip) {
+      for (int column = 0; static_cast<std::size_t>(column) < column_readers_.size();
+           ++column) {
+        SkipColumnRows(column, num_rows_remaining_to_skip);
+      }
+      current_row_ += num_rows_remaining_to_skip;
+      num_rows_remaining_to_skip = 0;
+    } else {
+      num_rows_remaining_to_skip -= num_rows_remaining_in_row_group;
+      current_row_ += num_rows_remaining_in_row_group;
+      NextRowGroup();
+    }
+  }
+  return num_rows_to_skip - num_rows_remaining_to_skip;
+}
+
+int64_t StreamReader::SkipColumns(int64_t num_columns_to_skip) {
+  int64_t num_columns_skipped = 0;
+
+  if (!eof_) {
+    for (; (num_columns_to_skip > num_columns_skipped) &&
+           static_cast<std::size_t>(column_index_) < nodes_.size();
+         ++column_index_) {
+      SkipColumnRows(column_index_, 1);
+      ++num_columns_skipped;
+    }
+  }
+  return num_columns_skipped;
+}
+
+void StreamReader::SkipColumnRows(int column, int64_t num_rows_to_skip) {
+  auto reader = column_readers_[column].get();
+  int64_t num_skipped = 0;
+
+  switch (reader->type()) {
+    case Type::BOOLEAN:
+      num_skipped = static_cast<BoolReader*>(reader)->Skip(num_rows_to_skip);
```

Comment:
is the cast necessary here?  I would expect this to be able to use virtual dispatch since it doesn't seem performance critical.

Reply:
Casting is necessary because the Skip() function is only available in the "TypedColumnReader" template class and not in the ColumnReader parent class.